File: DevEnvironment/DataStructures/src/main/java/datastructures/bitarray/core/Bit32ArrayReverseIterator.java
package datastructures.bitarray.core;

import base.core.Conditions;
import base.interfaces.IReverseIterator;
import datastructures.bitarray.interfaces.IBit32Array;

/**
 * The Bit32ArrayReverseIterator class implements a reverse iterator of a bit 32 array.
 */
public final class Bit32ArrayReverseIterator implements IReverseIterator<Boolean> {
    private final IBit32Array bitArray;
    private int currIndex;

    /**
     * Creates an iterator of a bit 32 array.
     */
    public static IReverseIterator<Boolean> make(IBit32Array bitArray) {
        return new Bit32ArrayReverseIterator(bitArray);
    }

    /**
     * The BitArrayIterator constructor.
     */
    private Bit32ArrayReverseIterator(IBit32Array bitArray) {
        Conditions.validateNotNull(
            bitArray,
            "The bit 32 array.");

        this.bitArray = bitArray;

        this.reset();
    }

    /**
     * Checks whether there is a next element.
     */
    @Override
    public boolean hasNext() {
        return this.currIndex >= 0;
    }

    /**
     * Gets the next element.
     */
    @Override
    public Boolean next() {
        assert(this.hasNext());

        boolean currElement = this.bitArray.isOn(this.currIndex);
        --this.currIndex;

        return currElement;
    }

    /**
     * Resets the iterator.
     */
    @Override
    public void reset() {
        this.currIndex = this.bitArray.size() - 1;
    }
}

File: SystemCore/DataStructures/bitarray/core/Bit32ArrayReverseIterator.cpp
#include "Bit32ArrayReverseIterator.h"
#include "Bit32Array.h"
#include <assert.h>

using namespace datastructures::bitarray;

/**
 * Creates a reverse iterator of a bit 32 array.
 */
base::IReverseIteratorSharedPtr<bool> Bit32ArrayReverseIterator::make(const IBit32Array& bitArray)
{
    return std::make_shared<Bit32ArrayReverseIterator>(bitArray);
}

/**
 * The Bit32ArrayReverseIterator constructor.
 */
Bit32ArrayReverseIterator::Bit32ArrayReverseIterator(const IBit32Array& bitArray) :
    m_bitArray(bitArray)
{
    Reset();
}

/**
 * The Bit32ArrayReverseIterator destructor.
 */
Bit32ArrayReverseIterator::~Bit32ArrayReverseIterator()
{
}

/**
 * Checks whether there is a Next element.
 */
bool Bit32ArrayReverseIterator::HasNext() const
{
    return !m_end;
}

/**
 * Gets the Next element.
 */
bool Bit32ArrayReverseIterator::Next()
{
    assert(HasNext());

    bool currElement = m_bitArray.IsOn(m_currIndex);

    if (m_currIndex == 0)
    {
        m_end = true;
    }
    else
    {
        --m_currIndex;
    }

    return currElement;
}

/**
 * Resets the iterator.
 */
void Bit32ArrayReverseIterator::Reset()
{
    m_currIndex = m_bitArray.Size() - 1;
    m_end = false;
}
